### PR TITLE
Api client lib

### PR DESCRIPTION
**Pull request by calavera**
:warning: Too big to review, go to my fork and navigate the api/client package :warning: 

I extracted a standalone client library from the cli that can be used outside of docker. In the process of doing that I end up removing a bunch of code from the cli.
## Why do you do this to us??!!

We should own the main go library to interact with docker. This makes the docker cli be the first consumer of this library.
## How did you do this??!

I extracted all the calls the client needs into an interface, you can see it here:
https://github.com/calavera/docker/blob/api_client_lib/api/client/client.go#L19

Functions are very well defined but there are a couple things to consider.
### I removed `cli.stream` and `cli.hijack` entirely.

After reading every single line in the client, I believe nobody knows how those two functions work and they made our code very complicated. You can compare the start command, no need for a diff:

Master: https://github.com/docker/docker/blob/master/api/client/start.go
My branch: https://github.com/calavera/docker/blob/api_client_lib/api/client/start.go

Every response in the new library that can be streamed returns an `io.ReadCloser`. It's up to the caller to implement streaming. This makes the intention of the caller much more clear. Compare the image create functions:

Master: https://github.com/docker/docker/blob/master/api/client/create.go#L53-L70
My branch: https://github.com/calavera/docker/blob/api_client_lib/api/client/create.go#L51-L63

did you know that the response of that function was always a json message that we stream in the client?
## Breaking changes

Of course there are always consequences. I introduced what can be considered as a breaking change.  In my opinion, the new behavior is more correct :smile_cat: 

Until now, `docker inspect` continued evaluating containers if you typed a wrong id, see this example:

```
$ docker run -d --name test alpine true
$ docker inspect this_container_does_not_exist test
```

The message about `this_container_does_not_exist` not existing would be followed by the test container output, a json payload of about 100 lines. This makes almost impossible to see the error.

With the new behavior, the cli stops execution as soon as it finds an error. This is a more predictable behavior. Users know sooner and more clear that there is a problem with the command they just executed.

The code in inspect has also been highly simplified. Compare both files:

Master: https://github.com/docker/docker/blob/master/api/client/inspect.go
My branch: https://github.com/calavera/docker/blob/api_client_lib/api/client/inspect.go
                   https://github.com/calavera/docker/blob/api_client_lib/api/client/inspect/inspector.go
## This is impossible to review

I'm afraid so too. I tried very hard to keep changes very well defined. But the best way to review this change might be to browse my branch directly, and to play with the cli.

The new client library is in https://github.com/calavera/docker/blob/api_client_lib/api/client/lib

### CODE REVIEW

**Comment by LK4D4** [comment]
I think probably inspect should fail if any of containers doesn't exist :)

**Comment by icecrime** [comment]
Oh my :heart_eyes: That looks pretty impressive!

**Comment by LK4D4** [comment]
Everything failed :fireworks:

**Comment by LK4D4** [comment]
Haha, seems like tests fixing won't be easy

**Comment by duglin** [comment]
Its kind of disturbing that the change in semantics you mentioned didn't result in a change to a testcase.  Aside from squashing this down :-)  can we get a new testcase for that?

btw - I really like the idea of a cli lib!

**Comment by vdemeester** [comment]
:heart_eyes_cat: Will have to take a look but `\o/`.

File: api/client/client.go [hunk 1]
```diff
@@ -3,3 +3,74 @@
...
+	ContainerStart(containerID string) error
+	ContainerStop(containerID string, timeout int) error
+	ContainerTop(containerID string, arguments []string) (types.ContainerProcessList, error)
+	ContainerUnpause(containerID string) error
+	ContainerWait(containerID string) (int, error)
+	CopyFromContainer(containerID, srcPath string) (io.ReadCloser, types.ContainerPathStat, error)
+	CopyToContainer(options types.CopyToContainerOptions) error
+	Events(options types.EventsOptions) (io.ReadCloser, error)
+	ImageBuild(options types.ImageBuildOptions) (types.ImageBuildResponse, error)
+	ImageCreate(options types.ImageCreateOptions) (io.ReadCloser, error)
```
Comment (tonistiigi) [inline api/client/client.go]: Do we need both `ImageCreate` and `ImagePull`? Their implementation looks the same and `ImageCreate` is only called by pull helper. Also `Parent` in `ImageCreate` looks misleading as it is still just a reference to pull(together with tag).

File: api/client/client.go [hunk 2]
```diff
@@ -3,3 +3,74 @@
...
+	ContainerAttach(options types.ContainerAttachOptions) (*types.HijackedResponse, error)
+	ContainerCommit(options types.ContainerCommitOptions) (types.ContainerCommitResponse, error)
+	ContainerCreate(config *runconfig.ContainerConfigWrapper, containerName string) (types.ContainerCreateResponse, error)
+	ContainerDiff(containerID string) ([]types.ContainerChange, error)
+	ContainerExecAttach(execID string, config *runconfig.ExecConfig) (*types.HijackedResponse, error)
+	ContainerExecCreate(config *runconfig.ExecConfig) (types.ContainerExecCreateResponse, error)
+	ContainerExecInspect(execID string) (types.ContainerExecInspect, error)
+	ContainerExecResize(execID string, height, width int) error
+	ContainerExecStart(execID string, config *types.ExecStartCheck) error
+	ContainerExport(containerID string) (io.ReadCloser, error)
```
Comment (tonistiigi) [inline api/client/client.go]: Lots of places where `containerID`/`imageID` is used it is not actually ID but a reference/prefix or container name. I think we should be clear about it on API level.

Comment (calavera) [inline api/client/client.go]: I know that. I personally like `containerID` as a reference more than `containerIDOrNameOrShortID`, I've tried to be consistent everywhere, but I'm open to suggestions.

File: api/client/client.go (hunk 1 above)
Comment (calavera) [inline api/client/client.go]: we probably don't need both, it's something to consider. I agree that the implementation is very similar.

File: api/client/client.go (hunk 2 above)
Comment (tonistiigi) [inline api/client/client.go]: How about just `container/image` or `containerRef/imageRef`?

File: api/client/client.go [hunk 3]
```diff
@@ -3,3 +3,74 @@
...
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/cliconfig"
+	"github.com/docker/docker/pkg/parsers/filters"
+	"github.com/docker/docker/registry"
+	"github.com/docker/docker/runconfig"
+)
+
+// apiClient is an interface that clients that talk with a docker server must implement.
+type apiClient interface {
+	ContainerAttach(options types.ContainerAttachOptions) (*types.HijackedResponse, error)
```
Comment (tonistiigi) [inline api/client/client.go]: Does this have to return pointer? + few other places (ExecStartCheck, ExecConfig). I think if there is no strict requirement for it we should be consistent with other methods.

File: api/client/stats.go [hunk 4]
```diff
@@ -156,18 +148,13 @@ func (cli *DockerCli) CmdStats(args ...string) error {
 	showAll := len(names) == 0
 
 	if showAll {
-		v := url.Values{}
-		if *all {
-			v.Set("all", "1")
+		options := types.ContainerListOptions{
+			All: true,
```
Comment (tonistiigi) [inline api/client/stats.go]: should still check for `all` flag

**Comment by tonistiigi** [comment]
`docker cp` seems to be broken. I get `http: read on closed response body` when copying from container and `Error response from daemon: Untar re-exec error: exit status 1: output: unexpected EOF` when copying to container. May may need to add more tests if this is the case(maybe they only check API atm).

**Comment by calavera** [comment]
@tonistiigi yep, there are still a few tests failing that I'm currently fixing. `docker cp` included.

File: api/client/stats.go (hunk 4 above)
Comment (calavera) [inline api/client/stats.go]: good catch! thank you.

**Comment by tianon** [comment]
Definitely high-level +1! :heart:

**Comment by jfrazelle** [comment]
nice love the description definitely helps :)

**Comment by calavera** [comment]
@duglin I totally agree with you, I was very suspicious about not seen inspect test failing. I've added some in my last commit. All inspects commands are implemented exactly in the same way, so all of them behave alike :raised_hands: 

https://github.com/docker/docker/commit/1db5748a35eb984a2e52ddb04a13723af0e7cbf9

File: api/client/client.go (hunk 2 above)
Comment (calavera) [inline api/client/client.go]: I like names without the `ID` suffix, but I'm going to be very consistent, which requires to change pretty much every new type in the library. So these are the variable names without it:

```
containerID => container
imageID     => image
networkID   => network
volumeID    => volume
execID      => exec
```

I'll make the necessary changes if we agree on those.

Comment (tonistiigi) [inline api/client/client.go]: I think they should be only replaced if they have the general meaning of `any-ref-to-container` or `any-ref-to-image`. For example, if create/commit return only IDs they should keep the `ID` suffix. Can `execID` be anything else than ID?

File: api/client/client.go (hunk 3 above)
Comment (calavera) [inline api/client/client.go]: I've changed this reference to be consistent.

**Comment by tiborvass** [comment]
@calavera don't really want to start a name bikeshedding, but I don't think it's a good idea to have a package be named `lib`. `dockerclient` would be more appropriate, plus it's also what was used for `samalba/dockerclient`.

**Comment by calavera** [comment]
@tiborvass that name is totally provisional, until we move all that code to a dedicated repo.

I am not going to call it `dockerclient` because it doesn't follow that other library's conventions and it can be confusing and misleading.

PS: The repo is already created, but it's private for now :smile_cat:

**Comment by calavera** [comment]
@tiborvass 

> I wonder if we shouldn't bring runconfig to the API

We should. I'd rather leave the configuration structs where they are for now, this is already big as it is. But we're working on consolidating those structs that are shared by client and server.

> I don't like that JSONConfig is a string

Fair enough, I'll make the change.

**Comment by tonistiigi** [comment]
I looked over most of the changes and played with this a bit. I think the API we provide should be in a bit higher level and easier to use for the third party apps, but I think there is no need to start going over these changes now. As a ported state of what our CLI needs this is a good starting point. But we should add a comment to the api package about this API being highly unstable atm.

I think we should try to get this in ASAP, to avoid heavy rebases and have plenty of time to test this before release. Everything doesn't need to be perfect in the first PR.

LGTM

**Comment by tiborvass** [comment]
Thanks for the fix, LGTM

**Comment by tiborvass** [comment]
userns had a flaky test I reported. Merging

**Comment by tiborvass** [comment]
My thoughts go to all the people who now have to rebase

**Comment by icecrime** [comment]
And to their families.

**Comment by thaJeztah** [comment]
And to those having to ping "needs a rebase", lol

**Comment by vdemeester** [comment]
:cat:

**Comment by jhowardmsft** [comment]
@calavera Can you take a look at this change - it's broken Windows building on master....

**Comment by calavera** [comment]
@jhowardmsft checking

**Comment by calavera** [comment]
@jhowardmsft it looks like windows is still building with Go 1.4.2 while the rest is building with 1.5. Is that intentional?

The problem is that this function was added to 1.5:

https://golang.org/pkg/text/template/#Template.Option

We can go back to the workaround we had with 1.4.2. Although I'd rather not, the fix for 1.4.2 is to replace the current rawElementInspect implementation with this function:

https://github.com/docker/docker/blob/release/v1.9/api/client/inspect.go#L175-L203

**Comment by duglin** [comment]
well, if we did go back to 1.4.2 that would fix the issues we're seeing with building on Power with gccgo where 1.5 isn't official available yet

ping @brahmaroutu

**Comment by tianon** [comment]
https://github.com/docker/docker/blob/master/project/PACKAGERS.md#build-dependencies

We do still officially support compiling with Go 1.4, so we'll need to
figure out something (or decide whether that requirement can be bumped, but
it's sounding like that's not really plausible right at the moment).

**Comment by calavera** [comment]
okay, I'll remove that call and go back to the ugly workaround :pensive: 

Thanks for checking!

**Comment by LK4D4** [comment]
@calavera You can use version build tags!

**Comment by calavera** [comment]
@LK4D4 indeed. Doing that.

**Comment by jhowardmsft** [comment]
@calavera Hang fire - I'll update the CI nodes, shouldn't take too long. I have already updated node 1 as I'm playing with Windows CI on it. @mikedougherty I assume you're OK me updating 2..9?

**Comment by brahmaroutu** [comment]
gccgo master builds failed because of this. I think the best would be to add gccgo PR builds to catch go1.5 related breakage earlier? 
@tianon go1.5 support is on main branch, can we build gcc:6.0 image from master and use it for x86, Power and Z?

**Comment by tianon** [comment]
It's OK, he's adding a shim for Go 1.4 support (which is still worthwhile
to keep for now IMO).

**Comment by jhowardmsft** [comment]
We have to move to 1.5.2 on those machines anyway. I just accelerated the update. All nodes are updated with the exception of Jenkins-prs-3. @mikedougherty I sent you an email about that node - I was unable to logon to it, even after a restart through the portal.

**Comment by WeiZhang555** [comment]
Wow, this is really amazing!! :+1:

**Comment by mikedougherty** [comment]
@jhowardmsft just got host 3 back up and running. sent you an email with details. thanks for the ping

**Pull request merged**